pkg/services/pluginsintegration/clientmiddleware/metrics_middleware.go

```diff
@@ -152,6 +173,32 @@ func (m *MetricsMiddleware) QueryData(ctx context.Context, req *backend.QueryDat
 	var resp *backend.QueryDataResponse
 	err := m.instrumentPluginRequest(ctx, req.PluginContext, endpointQueryData, func(ctx context.Context) (innerErr error) {
 		resp, innerErr = m.next.QueryData(ctx, req)
+		if resp == nil || resp.Responses == nil || !m.features.IsEnabled(featuremgmt.FlagPluginsInstrumentationStatusSource) {
+			return innerErr
+		}
+
+		// Set downstream status source in the context if there's at least one response with downstream status source,
+		// and if there's no plugin error
+		var hasPluginError bool
+		var hasDownstreamError bool
+		for _, r := range resp.Responses {
+			if r.Error == nil {
+				continue
+			}
+			if r.ErrorSource == backend.ErrorSourceDownstream {
+				hasDownstreamError = true
+			} else {
+				hasPluginError = true
+			}
+		}
+		// A plugin error has higher priority than a downstream error,
+		// so set to downstream only if there's no plugin error
+		if hasDownstreamError && !hasPluginError {
```

Comment:
curious as to why we have to set it to downstream here,  but we don't have to set it to plugin if this if condition is false

Reply:
I vote for 1) as well

Reply:
I have just implemented option 1